### ISSUE TITLE
Build with HPX_WITH_EXAMPLES fails  

### ISSUE THREAD

**Issue by scheherzade** [STEllAR-GROUP/hpx#5035]
HPX build falis when ```HPX_WITH_EXAMPLES``` is on.

## Actual Behavior
This is the error message I get:
```
Scanning dependencies of target version
[ 65%] Building CXX object libs/full/resiliency/examples/CMakeFiles/dataflow_replicate.dir/dataflow_replicate.cpp.o
[ 65%] Building CXX object libs/full/resiliency/examples/CMakeFiles/version.dir/version.cpp.o
[ 65%] Linking CXX executable ../../../../bin/custom_syntax
[ 65%] Linking CXX executable ../../../../bin/env_options
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/custom_syntax.dir/build.make:114: bin/custom_syntax] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42348: libs/full/program_options/examples/CMakeFiles/custom_syntax.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 65%] Linking CXX executable ../../../../bin/real
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/env_options.dir/build.make:114: bin/env_options] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42522: libs/full/program_options/examples/CMakeFiles/env_options.dir/all] Error 2
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/real.dir/build.make:114: bin/real] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42553: libs/full/program_options/examples/CMakeFiles/real.dir/all] Error 2
[ 65%] Linking CXX executable ../../../../bin/options_description
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/options_description.dir/build.make:114: bin/options_description] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42882: libs/full/program_options/examples/CMakeFiles/options_description.dir/all] Error 2
[ 65%] Linking CXX executable ../../../../bin/config_file_types
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/config_file_types.dir/build.make:114: bin/config_file_types] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42729: libs/full/program_options/examples/CMakeFiles/config_file_types.dir/all] Error 2
[ 65%] Linking CXX shared library ../../../lib/libhpx_jacobi.so
[ 65%] Built target jacobi_component
[ 65%] Linking CXX executable ../../../../bin/response_file
[ 68%] Linking CXX executable ../../../../bin/options_hierarchy
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/response_file.dir/build.make:114: bin/response_file] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42124: libs/full/program_options/examples/CMakeFiles/response_file.dir/all] Error 2
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/options_hierarchy.dir/build.make:114: bin/options_hierarchy] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42182: libs/full/program_options/examples/CMakeFiles/options_hierarchy.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/option_groups
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/option_groups.dir/build.make:114: bin/option_groups] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42460: libs/full/program_options/examples/CMakeFiles/option_groups.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/multiple_sources
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
[ 68%] Linking CXX executable ../../../../bin/first
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/multiple_sources.dir/build.make:114: bin/multiple_sources] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42967: libs/full/program_options/examples/CMakeFiles/multiple_sources.dir/all] Error 2
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/first.dir/build.make:114: bin/first] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42824: libs/full/program_options/examples/CMakeFiles/first.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/regex
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/program_options/examples/CMakeFiles/regex.dir/build.make:114: bin/regex] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:42491: libs/full/program_options/examples/CMakeFiles/regex.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/async_replay
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/resiliency/examples/CMakeFiles/async_replay.dir/build.make:114: bin/async_replay] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:44184: libs/full/resiliency/examples/CMakeFiles/async_replay.dir/all] Error 2
[ 68%] Linking CXX shared library ../../lib/libhpx_component_storage.so
[ 68%] Built target component_storage_component
[ 68%] Linking CXX executable ../../../../bin/version
[ 68%] Linking CXX executable ../../../../bin/async_replicate_vote
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/resiliency/examples/CMakeFiles/version.dir/build.make:114: bin/version] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:44787: libs/full/resiliency/examples/CMakeFiles/version.dir/all] Error 2
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/resiliency/examples/CMakeFiles/async_replicate_vote.dir/build.make:114: bin/async_replicate_vote] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:44559: libs/full/resiliency/examples/CMakeFiles/async_replicate_vote.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/1d_stencil_replay_exception
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/resiliency/examples/CMakeFiles/1d_stencil_replay_exception.dir/build.make:114: bin/1d_stencil_replay_exception] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:44497: libs/full/resiliency/examples/CMakeFiles/1d_stencil_replay_exception.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/1d_stencil_replay_validate
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status
gmake[2]: *** [libs/full/resiliency/examples/CMakeFiles/1d_stencil_replay_validate.dir/build.make:114: bin/1d_stencil_replay_validate] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:44528: libs/full/resiliency/examples/CMakeFiles/1d_stencil_replay_validate.dir/all] Error 2
[ 68%] Linking CXX executable ../../../../bin/1d_stencil_4_checkpoint
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
/usr/bin/ld: ../../../../lib/libhpx.so.1.6.0: undefined reference to `bool hpx::detail::apply_impl<hpx::lcos::base_lco_with_value<hpx::naming::id_type, hpx::naming::gid_type, hpx::traits::detail::managed_component_tag>::set_value_action, hpx::naming::gid_type>(hpx::naming::id_type const&, hpx::naming::address&&, hpx::threads::thread_priority, hpx::naming::gid_type&&)'
collect2: error: ld returned 1 exit status

```

## Steps to Reproduce the Problem
Build HPX without setting any specific flags.

## Specifications
  - HPX Version: master
  - Platform (compiler, OS): Linux, Fedora32, GCC 10.2.1

```
-- The CXX compiler identification is GNU 10.2.1
-- The C compiler identification is GNU 10.2.1
-- Check for working CXX compiler: /usr/lib64/ccache/c++
-- Check for working CXX compiler: /usr/lib64/ccache/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/lib64/ccache/cc
-- Check for working C compiler: /usr/lib64/ccache/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Using C++14
-- CMake version: 3.17.3
-- HPX version: 1.6.0
-- GIT commit is 8f608a97ae4f765cb8fdd166b478097267d65c4b.
```

**Comment by msimberg** [comment]
This is most likely specific to newer GCCs. Our gcc-newest builder (with GCC 10) has the distributed runtime off, so these examples don't get built. gcc-oldest (with GCC 7) has examples enabled but these build fine there. All versions of clang seem to be OK. @hkaiser do you already have a suspect or would you need help to reproduce/fix this?

**Comment by hkaiser** [comment]
> This is most likely specific to newer GCCs. Our gcc-newest builder (with GCC 10) has the distributed runtime off, so these examples don't get built. gcc-oldest (with GCC 7) has examples enabled but these build fine there. All versions of clang seem to be OK. @hkaiser do you already have a suspect or would you need help to reproduce/fix this?

@msimberg I don't have a suspect ATM, so any help you can provide would be great. In the end it's most probably just a missing header #include somewhere.

**Comment by scheherzade** [comment]
PR [Fix GCC 10 release builds #5039](https://github.com/STEllAR-GROUP/hpx/pull/5039) fixes it, thanks!
